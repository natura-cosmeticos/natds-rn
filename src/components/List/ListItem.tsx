import React, { useState, ReactElement } from 'react';
import { LayoutChangeEvent, View } from 'react-native';
import { Theme } from '../../common/themeSelectors';
import { TouchableRipple, TouchableRippleProps } from '../TouchableRipple/TouchableRipple';
import { ListItem as ListItemComponent } from './ListItem.styles';

export interface ListItemProps {
  /**
   * The children list item elements
   */
  children: ReactElement;
  /**
   * Controls the element selected styles
   */
  selected?: boolean;
  /**
   * onPress event handler
   */
  onPress?(): void;
  /**
   * Id for testing
   */
  testID?: string;
  /**
   * App's theme
   */
  theme?: Theme;
}

export const getRippleSize = (
  event: LayoutChangeEvent,
  setSize: React.Dispatch<React.SetStateAction<number>>,
) => {
  const { layout: { width, height } } = event.nativeEvent;
  const biggerSide = width >= height ? width : height;

  setSize((biggerSide / 2));
};

const ListWrapper = ({ onPress, children, ...props }: TouchableRippleProps) => {
  if (onPress) {
    return (
      <TouchableRipple onPress={onPress} {...props}>
        {children}
      </TouchableRipple>
    );
  }

  return (
    <View>
      {children}
    </View>
  );
};

export const ListItem = ({
  children,
  onPress,
  testID = 'list-item',
  ...rest
}: ListItemProps) => {
  const [size, setSize] = useState(0);

  return (
    <ListWrapper
      color="highlight"
      hideOverflow
      size={size}
      onPress={onPress}
      testID={`${testID}-wrapper`}
    >
      <ListItemComponent
        testID={testID}
        onLayout={event => onPress && getRippleSize(event, setSize)}
        onPress={onPress}
        {...rest}
      >
        {children}
      </ListItemComponent >
    </ListWrapper >
  );
};
