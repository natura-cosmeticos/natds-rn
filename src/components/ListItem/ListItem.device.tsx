import { storiesOf } from '@storybook/react-native'
import {
  Default,
  Dividers,
  RippleFeedback,
  SelectionFeedback
} from './ListItem.stories'

storiesOf('ListItem', module)
  .add('Default', Default)
  .add('FeedbackRipple', RippleFeedback)
  .add('FeedbackSelection', SelectionFeedback)
  .add('Dividers', Dividers)
