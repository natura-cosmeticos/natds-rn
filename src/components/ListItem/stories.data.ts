export const data = [
  {
    key: 'item1',
    title: 'A Game of Thrones ',
  },
  {
    key: 'item2',
    title: 'A Clash of Kings',
  },
  {
    key: 'item3',
    title: 'A Storm of Swords',
  },
  {
    key: 'item4',
    title: 'A Feast for Crows',
  },
  {
    extraInfo: {
      author: 'George R.R. Martin',
      rating: 10,
      releaseDate: '2000',
    },
    key: 'item5',
    title: 'A Dance with Dragons',
  },
  {
    key: 'item6',
    title: 'The Winds of Winter',
    unreleased: true,
  },
  {
    key: 'item7',
    title: 'A Dream of Spring',
    unreleased: true,
  },
];
