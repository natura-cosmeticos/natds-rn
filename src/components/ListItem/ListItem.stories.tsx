/* eslint-disable max-lines */
/* eslint-disable no-console */
import React, { useState } from 'react'
import { View, FlatList, Text } from 'react-native'
import { Card } from '../Card'
import { Divider } from '../Divider'
import { ListItem, ListItemFeedback } from './ListItem'
import { Checkbox } from '../Checkbox'
import { TextWithTheme } from '../../common/HelperComponents/ThemeHelper.styles'
import { data } from './stories.data'
import { StoryContainer } from '../../common/HelperComponents/StoryContainer'

export default {
  component: ListItem,
  parameters: {
    componentSubtitle: 'List items are rows of information that can have user interactions or not'
  },
  title: 'Components|List'
}

export const Default = () => (
  <StoryContainer title="Standard">
    <FlatList
      data={data}
      renderItem={({ item }) => (
        <ListItem>
          <View style={{ padding: 16 }}>
            <TextWithTheme>{item.title}</TextWithTheme>
          </View>
        </ListItem>
      )}
    />
  </StoryContainer>
)

export const RippleFeedback = () => (
  <StoryContainer title="Feedback Ripple">
    <FlatList
      data={data}
      renderItem={({ item }) => (
        <ListItem feedback="ripple" onPress={() => console.log('I was pressed')}>
          <View style={{ padding: 16 }}>
            <TextWithTheme>{item.title}</TextWithTheme>
          </View>
        </ListItem>
      )}
    />
  </StoryContainer>
)

export const SelectionFeedback = () => {
  const [selected, setSelected] = useState('')

  const renderItem = ({ item }) => {
    const listItemProps = {
      feedback: 'selection' as ListItemFeedback,
      selected: item.key === selected,
      ...(item.unreleased ? {} : { onPress: () => setSelected(item.key) })
    }

    return (
      <ListItem {...listItemProps}>
        <View style={{
          flexDirection: 'row',
          justifyContent: 'space-between',
          padding: 16
        }}
        >
          <TextWithTheme>{item.title}</TextWithTheme>
          <Checkbox
            selected={item.key === selected}
            onPress={() => setSelected(item.key)}
            disabled={item.unreleased}
          />
        </View>
      </ListItem>
    )
  }

  return (
    <StoryContainer title="Feedback Selection">
      <FlatList
        ItemSeparatorComponent={Divider}
        data={data}
        renderItem={renderItem}
        extraData={selected}
      />
    </StoryContainer>
  )
}

export const Dividers = () => {
  const renderExtraInfo = (extraInfo) => {
    if (!extraInfo) return null

    const extraInfoData = Object
      .keys(extraInfo)
      .map((item, index) => (
        <View key={extraInfo[item].key} style={{ backgroundColor: '#fafafa' }}>
          <ListItem>
            <View style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              paddingHorizontal: 32,
              paddingVertical: 16
            }}
            >
              <TextWithTheme>{item}</TextWithTheme>
              <TextWithTheme>{extraInfo[item]}</TextWithTheme>
            </View>
          </ListItem>
          {
            index < Object.keys(extraInfo).length - 1
              ? <Divider type="inset" />
              : null
          }
        </View>
      ))

    return extraInfoData
  }

  return (
    <StoryContainer title="Dividers">
      <View style={{
        flex: 1,
        paddingHorizontal: 10,
        paddingVertical: 20
      }}
      >
        <Card>
          <FlatList
            ItemSeparatorComponent={Divider}
            data={data}
            renderItem={({ item }) => (
              <>
                <ListItem onPress={() => console.log('I was pressed')}>
                  <View style={{ padding: 16 }}>
                    <Text>{item.title}</Text>
                  </View>
                </ListItem>
                {renderExtraInfo(item.extraInfo)}
              </>
            )}
          />
        </Card>
      </View>
    </StoryContainer>
  )
}
