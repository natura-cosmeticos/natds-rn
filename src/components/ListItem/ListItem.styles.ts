import styled from 'styled-components/native'
import {
  getColorPrimary,
  getOpacityMediumLow,
  buildColorWithOpacity,
  getSize,
  getTheme
} from '../../common/themeSelectors'
import { ListItemProps } from './ListItem'

const backgroundStyles = ({ selected, theme }: Required<ListItemProps>) => (
  selected
    ? buildColorWithOpacity(getColorPrimary, getOpacityMediumLow, theme)
    : 'transparent'
)

const getSizeMedium = getTheme((theme) => getSize(theme, 'medium'))

export const ListItem = styled.View<ListItemProps>`
  background: ${backgroundStyles};
  min-height: ${getSizeMedium}px;
`
