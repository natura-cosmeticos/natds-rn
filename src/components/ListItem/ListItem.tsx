import React, { ReactElement } from 'react'
import { View, TouchableWithoutFeedback } from 'react-native'
import { Theme } from '../../common/themeSelectors'
import { TouchableRipple, TouchableRippleProps } from '../TouchableRipple/TouchableRipple'
import { ListItem as ListItemComponent } from './ListItem.styles'

export type ListItemFeedback = 'ripple' | 'selection'

export interface ListItemProps {
  /**
   * The children list item elements
   */
  children: ReactElement;
  /**
   * Controls the element feedback type `ripple` | `selection`
   * @Default ripple
   */
  feedback?: ListItemFeedback;
  /**
   * Controls the element selected styles when feedback type is `selection`
   * @Default false
   */
  selected?: boolean;
  /**
   * onPress event handler
   */
  onPress?(): void;
  /**
   * Id for testing
   */
  testID?: string;
  /**
   * App's theme
   */
  theme?: Theme;
}

type ListWrapper = TouchableRippleProps & Pick<ListItemProps, 'feedback'>

const ListWrapper = ({
  onPress, children, feedback, ...props
}: ListWrapper) => {
  if (!onPress) {
    return (
      <View>
        {children}
      </View>
    )
  }

  if (feedback === 'ripple') {
    return (
      <TouchableRipple onPress={onPress} {...props}>
        {children}
      </TouchableRipple>
    )
  }

  return (
    <TouchableWithoutFeedback onPress={onPress} {...props}>
      {children}
    </TouchableWithoutFeedback>
  )
}

export const ListItem = ({
  children,
  onPress,
  testID = 'list-item',
  feedback = 'ripple',
  ...rest
}: ListItemProps) => (
  <ListWrapper
    color="highlight"
    hideOverflow
    onPress={onPress}
    testID={`${testID}-wrapper`}
    feedback={feedback}
  >
    <ListItemComponent
      testID={testID}
      onPress={onPress}
      {...rest}
    >
      {children}
    </ListItemComponent>
  </ListWrapper>
)
