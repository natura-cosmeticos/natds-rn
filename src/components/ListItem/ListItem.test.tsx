import React from 'react'
import { Text } from 'react-native'
import { fireEvent } from '@testing-library/react-native'
import { renderWithTheme } from '../../../test/testHelpers'
import { ListItem, ListItemProps } from './ListItem'

const renderList = (props?: Omit<ListItemProps, 'children'>) => (
  renderWithTheme(
    <ListItem {...props}>
      <Text>Text</Text>
    </ListItem>
  )
)

describe('ListItem component', () => {
  it('Should render ListItem component correctly', () => {
    const { toJSON } = renderList()

    expect(toJSON()).toMatchSnapshot()
  })

  it('Should render ListItem with state selected', () => {
    const { toJSON } = renderList({ selected: true })

    expect(toJSON()).toMatchSnapshot()
  })

  it('Should render ListItem with ripple if is clickable', () => {
    const { toJSON } = renderList({ onPress: jest.fn() })

    expect(toJSON()).toMatchSnapshot()
  })

  it('Should not render ListItem with ripple if is selected', () => {
    const { toJSON } = renderList({ onPress: jest.fn(), selected: true })

    expect(toJSON()).toMatchSnapshot()
  })

  it('Should call the given onPress function', () => {
    const onPress = jest.fn()
    const { queryByTestId } = renderList({ onPress })

    const listItem = queryByTestId('list-item-wrapper')

    if (listItem) {
      fireEvent.press(listItem)
    }

    expect(onPress).toHaveBeenCalled()
  })
})
