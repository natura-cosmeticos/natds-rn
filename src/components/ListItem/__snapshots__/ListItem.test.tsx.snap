// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem component Should not render ListItem with ripple if is selected 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="list-item-wrapper"
>
  <View
    onPress={[MockFunction]}
    selected={true}
    style={
      Array [
        Object {
          "backgroundColor": "#f4ab3416",
          "minHeight": 48,
        },
      ]
    }
    testID="list-item"
  >
    <Text>
      Text
    </Text>
  </View>
</View>
`;

exports[`ListItem component Should render ListItem component correctly 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "minHeight": 48,
        },
      ]
    }
    testID="list-item"
  >
    <Text>
      Text
    </Text>
  </View>
</View>
`;

exports[`ListItem component Should render ListItem with ripple if is clickable 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="list-item-wrapper"
>
  <View
    onPress={[MockFunction]}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "minHeight": 48,
        },
      ]
    }
    testID="list-item"
  >
    <Text>
      Text
    </Text>
  </View>
</View>
`;

exports[`ListItem component Should render ListItem with state selected 1`] = `
<View>
  <View
    selected={true}
    style={
      Array [
        Object {
          "backgroundColor": "#f4ab3416",
          "minHeight": 48,
        },
      ]
    }
    testID="list-item"
  >
    <Text>
      Text
    </Text>
  </View>
</View>
`;
